dask_cuda/explicit_comms/dataframe_shuffle.py

```diff
@@ -0,0 +1,260 @@
+import asyncio
+
+import pandas
+
+from dask.base import is_dask_collection
+from dask.dataframe.core import _Frame
+from dask.dataframe.shuffle import partitioning_index, shuffle_group
+from distributed.protocol import to_serialize
+
+from . import comms
+
+
+async def send_df(ep, df):
+    if df is None:
+        return await ep.write("empty")
+    else:
+        return await ep.write([to_serialize(df)])
+
+
+async def recv_df(ep):
+    ret = await ep.read()
+    if ret == "empty":
+        return None
+    else:
+        return ret[0]
+
+
+async def barrier(rank, eps):
+    if rank == 0:
+        await asyncio.gather(*[ep.read() for ep in eps.values()])
+    else:
+        await eps[0].write("dummy")
+
+
+async def send_bins(eps, bins):
+    futures = []
+    for rank, ep in eps.items():
+        futures.append(send_df(ep, bins[rank]))
+    await asyncio.gather(*futures)
+
+
+async def recv_bins(eps, bins):
+    futures = []
+    for ep in eps.values():
+        futures.append(recv_df(ep))
+    bins.extend(await asyncio.gather(*futures))
+
+
+async def exchange_and_concat_bins(rank, eps, bins):
+    ret = [bins[rank]]
+    await asyncio.gather(recv_bins(eps, ret), send_bins(eps, bins))
+    return concat([df for df in ret if df is not None])
+
+
+def concat(df_list):
```

Comment:
Can we use [`_concat`](https://github.com/dask/dask/blob/a6615bd175bec61ea62e7d5187ea7d6e2da2dee7/dask/dataframe/core.py#L83)?

Reply:
Done